pkgs/development/python-modules/pur/default.nix

```diff
@@ -1,19 +1,26 @@
 { lib
-, buildPythonPackage
-, click
+, python3
 , fetchFromGitHub
-, pytestCheckHook
 }:
 
+let
+  py = python3.override {
+    packageOverrides = self: super: {
+      click = self.callPackage ../../../development/python-modules/click/7.nix { };
+    };
+  };
+  inherit (py.pkgs) buildPythonPackage click pytestCheckHook;
+in
+
```

Comment:
Pinning within python-modules is highly discouraged. This will potentially introduce incompatible versions in a user's environment as either the new pinned version or the original version will be imported at runtime, breaking one of the packages.

The preference to handling this is to relax version bounds in the "install_requires" field. (could be in setup.py, pyproject.toml, requirements or others). In most cases, packages are still compatible with small API changes which may warrant a major version bump. We use test suites to verify that the package still works correctly. 

If the package is still incompatible with the latest major version, then the most proper way to handle this is make an issue with the upstream package to adopt the latest major version. Or if upstream is not very responsive, you are free patch the source to make it compatible.

In very few circumstances, two versions of the same package are allowed to exist if the packages are extremely difficult to package. Some examples of this are tensorflow, which has huge ecosystems built around it and is hard to package. Another is django, which has 2 actively developed versions, and large ecosystems built around each.

One exception to this is applications, due to the way `buildPythonApplication` and `toPythonApplication` functions work, the related derivations will not bleed dependencies between packages. If the package doesn't need to be imported by other python modules, then this package would be a good candidate to convert into application. You can look at https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/admin/awscli/default.nix as an example of using an overlay within a python application. 

Another valid exception is pinning within the same interpreter version. If a package no longer supports older-but-still-relevant interpreter versions, then you're free to pin a package for an interpreter. This was very common with python2, but less common amongst python3 interpreters. (e.g. `if (pythonOlder "3.8" then <older pkg> else <latest package> ;`)

Info on `buildPythonApplication` can be found [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#buildpythonapplication-function).
Info on `toPythonApplication` can be found [here](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#topythonapplication-function).

Reply:
Thank you for this detailed explanation! 
As I already said in the description, this python module does not work with click versions 8.0 or higher.
The maintainers are aware of this, and have promised to fix it but no progress has been made in the last two months. 

I only saw this package failing on hydra, and thought the fix was trivial, not knowing it could cause problems.

As a workaround, is what I suggested [here](https://github.com/NixOS/nixpkgs/pull/144955#discussion_r744192060) a valid alternative?

Reply:
> As a workaround, is what I suggested here a valid alternative?

Sadly not because of this:

> This is not allowed in pythonPackages because python cannot mix multiple versions of the same package and things break in really strange ways when it happens.

I think we either need to convert this package to use `buildPythonApplication` (seems acceptable, since looking at the package at [PyPI](https://pypi.org/project/pur/) it doesn't seem that is meant to be used as an library) or mark it as broken (liking at the PR explaining the click version issue).